codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/RuntimeClientPlugin.java

```diff
@@ -41,37 +41,39 @@
 public final class RuntimeClientPlugin implements ToSmithyBuilder<RuntimeClientPlugin> {
 
     private final Symbol resolveFunction;
-    private final ApplyPlugin buildMiddlewareStack;
+    private final ApplyPlugin writeMiddlewareHelpers;
     private final BiPredicate<Model, ServiceShape> servicePredicate;
     private final OperationPredicate operationPredicate;
     private final Set<ConfigField> configFields;
+    private final MiddlewareRegistrar registerMiddleware;
+
 
     private RuntimeClientPlugin(Builder builder) {
         resolveFunction = builder.resolveFunction;
         operationPredicate = builder.operationPredicate;
         servicePredicate = builder.servicePredicate;
         configFields = builder.configFields;
-        buildMiddlewareStack = builder.buildMiddlewareStack;
+        writeMiddlewareHelpers = builder.writeMiddlewareHelpers;
+        registerMiddleware = builder.registerMiddleware;
     }
 
     @FunctionalInterface
     public interface ApplyPlugin {
 
         /**
-         * applyMiddleware applies the middleware onto the Operation stack.
+         * generateMiddlewareHelper generates a middleware helper function
+         * for a service or operation shape.
          *
          * @param writer GoWriter to write the Go code.
          * @param service service shape to which the operation belongs.
          * @param operation operation shape for which operation stack is defined.
          * @param protocolGenerator Protocol generator for protocol used by service.
-         * @param stackOperand operand to denote the stack to which middleware is applied.
          */
-        void applyMiddleware(
+        void generateMiddlewareHelper(
```

Comment:
This function should not be needed. Static middleware doesn't need it and generated middleware should be doing generation via the hooks in `GoIntegration`.

Reply:
The reason for making this a RuntimeClient plugin instead of `GoIntegration` was that none of the existing GoIntegration methods take in a `ProtocolGenerator` which is required for knowing if valid serializer/deserializer middlewares are generated. 

We can try to override `writeAdditionalFiles` integration hooks but those will have to have redundant methods to retrieve protocol used as it doesn't take.  Alternatively can define a seperate plugin definition which would have to make this a special plugin.

Reply:
Why do you need the protocol generator? You can safely assume that serde will be generated. The only place I saw you using that was in actually registering the serde, which should be done directly in `OperationGenerator` anyway.